Source/CarthageKit/Project.swift

```diff
@@ -684,50 +698,85 @@ public final class Project {
+					.map { String($0) }
+					.flatMap(flatMapTransform)
+			}
+	}
+
 	/// Creates symlink between the dependency checkouts and the root checkouts
-	private func symlinkCheckoutPathsForDependencyProject(_ dependency: ProjectIdentifier, subDependencies: Set<ProjectIdentifier>, rootDirectoryURL: URL) -> SignalProducer<(), CarthageError> {
-		let rootCheckoutsURL = rootDirectoryURL.appendingPathComponent(CarthageProjectCheckoutsPath, isDirectory: true).resolvingSymlinksInPath()
-		let rawDependencyURL = rootDirectoryURL.appendingPathComponent(dependency.relativePath, isDirectory: true)
+	private func symlinkCheckoutPaths(for dependency: Dependency<PinnedVersion>, withRepository repositoryURL: URL, atRootDirectory rootDirectoryURL: URL) -> SignalProducer<(), CarthageError> {
+		let rawDependencyURL = rootDirectoryURL.appendingPathComponent(dependency.project.relativePath, isDirectory: true)
 		let dependencyURL = rawDependencyURL.resolvingSymlinksInPath()
 		let dependencyCheckoutsURL = dependencyURL.appendingPathComponent(CarthageProjectCheckoutsPath, isDirectory: true).resolvingSymlinksInPath()
-		let subDependencyNames = subDependencies.map { $0.name }
 		let fileManager = FileManager.default
 
-		let symlinksProducer = SignalProducer<String, CarthageError>(subDependencyNames)
-			.filter { name in
-				let checkoutURL = rootCheckoutsURL.appendingPathComponent(name)
-				do {
-					return try checkoutURL.resourceValues(forKeys: [ .isDirectoryKey ]).isDirectory ?? false
-				} catch {
-					return false
-				}
-			}
-			.attemptMap { name -> Result<(), CarthageError> in
-				let dependencyCheckoutURL = dependencyCheckoutsURL.appendingPathComponent(name)
-				let subdirectoryPath = (CarthageProjectCheckoutsPath as NSString).appendingPathComponent(name)
-				let linkDestinationPath = relativeLinkDestinationForDependencyProject(dependency, subdirectory: subdirectoryPath)
-				do {
-					try fileManager.createSymbolicLink(atPath: dependencyCheckoutURL.carthage_path, withDestinationPath: linkDestinationPath)
-				} catch let error as NSError {
-					if !(error.domain == NSCocoaErrorDomain && error.code == NSFileWriteFileExistsError) {
-						return .failure(.writeFailed(dependencyCheckoutURL, error))
+		return self.dependencyProjects(for: dependency)
+			.zip(with: // file system objects which might conflict with symlinks
+				fileSystemObjects(for: dependency, withRepository: repositoryURL, at: CarthageProjectCheckoutsPath, flatMapTransform: { path in
+					let componentsRelativeToDirectoryURL = {
+						return URL(string: $0, relativeTo: self.directoryURL)?.standardizedFileURL.carthage_pathComponents
 					}
-				}
-				return .success()
-			}
+					if
+						let components = componentsRelativeToDirectoryURL(path),
+						let comparator = componentsRelativeToDirectoryURL(CarthageProjectCheckoutsPath),
+						Array(components.dropLast(1)) == comparator
+					{
+						return components.last!
+					} else {
+						return nil
+					}
```

Comment:
I don't really understand what this section is doing. I think it could use a comment. (The before code was equally confusing. But I think you understand this now?)